File: Homework/Homework 2/C#/Exercise 1 - C#/Exercise 1/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Exercise_1
{
    public partial class Form1 : Form
    {
        
        public Random R = new Random();
        public Double CurrentAvarage = 0;
        public int countNumber = 0;

        public SortedDictionary<int, frequencies> FrequencyDistribution = new SortedDictionary<int, frequencies>();

        public Form1()
        {
            InitializeComponent();

            this.richTextBox1.AppendText("Exam".PadRight(10) + "Grade".PadRight(8) + "Current Mean" + Environment.NewLine);

            this.chart1.ChartAreas["ChartArea1"].AxisX.MinorTickMark.Enabled = true;
            this.chart1.ChartAreas["ChartArea1"].AxisX.Interval = 1;
            this.chart1.ChartAreas["ChartArea1"].AxisX.IsLabelAutoFit = true;

        }

        private void timer1_Tick(object sender, EventArgs e)
        {

            //Online arithmetical mean
            //----------------------------------------------------

            //Increment the counter value
            countNumber += 1;

            //The Knuth formula
            int grade = R.Next(18, 31);
            CurrentAvarage = CurrentAvarage + (grade - CurrentAvarage) / countNumber;

            string nameOfTheExam = "Exam " + countNumber;

            this.richTextBox1.AppendText(nameOfTheExam.PadRight(10) +
                                       grade.ToString().PadRight(8) + CurrentAvarage +
                                       Environment.NewLine);




            //---------------------------------------------------
            //Distribution

            if (FrequencyDistribution.ContainsKey(grade))
            {
                FrequencyDistribution[grade].countFrequencies += 1;
            }
            else
            {
                FrequencyDistribution.Add(grade, new frequencies());
            }

            //Clear every time
            this.richTextBox2.Clear();
            this.chart1.Series["Relative Frequency"].Points.Clear();
            this.richTextBox2.AppendText("Grade".PadRight(7) + "Count".PadRight(7) + "Freq".PadRight(7) + "Perc".PadRight(7) + Environment.NewLine);

            //Calculate the relative frequencies and percentage frequencies. Then draw the graphic
            foreach (KeyValuePair<int, frequencies> freq in FrequencyDistribution)
            {
                FrequencyDistribution[freq.Key].RelativeFrequencies = Convert.ToDouble(FrequencyDistribution[freq.Key].countFrequencies) / Convert.ToDouble(countNumber);
                FrequencyDistribution[freq.Key].PercentageFrequencies = Convert.ToDouble(FrequencyDistribution[freq.Key].RelativeFrequencies) * 100;

                string fr = "0." + freq.Value.RelativeFrequencies.ToString("0.##");
                this.chart1.Series["Relative Frequency"].Points.AddXY(freq.Key.ToString(), fr);

                this.richTextBox2.AppendText(freq.Key.ToString().PadRight(7) +
                                           freq.Value.countFrequencies.ToString().PadRight(7) +
                                           freq.Value.RelativeFrequencies.ToString("0.##").PadRight(7) +
                                           freq.Value.PercentageFrequencies.ToString("0.##").PadRight(7) + " % " + Environment.NewLine);



            }

            this.richTextBox2.AppendText(Environment.NewLine + Environment.NewLine +
                                       "Total Count: " + countNumber + Environment.NewLine);



        }

        private void button1_MouseClick(object sender, MouseEventArgs e)
        {
            this.timer1.Start();
        }

        private void button2_MouseClick(object sender, MouseEventArgs e)
        {
            this.timer1.Stop();
        }
    }

    // Class Frequencies
    public class frequencies
    {

        public int countFrequencies = 1;
        public Double RelativeFrequencies;
        public Double PercentageFrequencies;

    }
}

File: Homework/Homework 2/VB.NET/Exercise 1 - vb/Exercise 1 - vb/Form1.vb
Public Class Form1

    Public R As New Random
    Public CurrentAvarage As Double = 0
    Public countNumber As Integer = 0


    Dim FrequencyDistribution As New SortedDictionary(Of Integer, frequencies)


    Private Sub Timer1_Tick(sender As Object, e As EventArgs) Handles Timer1.Tick

        'Online arithmetical mean
        '----------------------------------------------------

        'Increment the counter value
        countNumber += 1

        'The Knuth formula
        Dim grade As Integer = Me.R.Next(18, 31)
        CurrentAvarage = CurrentAvarage + (grade - CurrentAvarage) / countNumber

        Dim nameOfTheExam = "Exam " & countNumber

        Me.RichTextBox1.AppendText(nameOfTheExam.PadRight(10) &
                                   grade.ToString.PadRight(8) & CurrentAvarage &
                                   Environment.NewLine)




        '---------------------------------------------------
        'Distribution

        If FrequencyDistribution.ContainsKey(grade) Then
            FrequencyDistribution(grade).countFrequencies += 1
        Else
            FrequencyDistribution.Add(grade, New frequencies)
        End If

        Me.RichTextBox2.Clear()
        Me.Chart1.Series("Relative Frequency").Points.Clear()
        Me.RichTextBox2.AppendText("Grade".PadRight(7) & "Count".PadRight(7) & "Freq".PadRight(7) & "Perc".PadRight(7) & Environment.NewLine)


        'Calculate the relative frequencies and percentage frequencies. Then draw the graphic
        For Each freq As KeyValuePair(Of Integer, frequencies) In FrequencyDistribution

            FrequencyDistribution(freq.Key).RelativeFrequencies = FrequencyDistribution(freq.Key).countFrequencies / countNumber
            FrequencyDistribution(freq.Key).PercentageFrequencies = FrequencyDistribution(freq.Key).RelativeFrequencies * 100

            Dim fr = "0." & freq.Value.RelativeFrequencies.ToString("0.##")
            Me.Chart1.Series("Relative Frequency").Points.AddXY(freq.Key.ToString, fr)

            Me.RichTextBox2.AppendText(freq.Key.ToString.PadRight(7) &
                                       freq.Value.countFrequencies.ToString.PadRight(7) &
                                       freq.Value.RelativeFrequencies.ToString("0.##").PadRight(7) &
                                       freq.Value.PercentageFrequencies.ToString("0.##").PadRight(7) & " % " & Environment.NewLine)



        Next

        Me.RichTextBox2.AppendText(Environment.NewLine & Environment.NewLine &
                                   "Total Count: " & countNumber & Environment.NewLine)





    End Sub

    Private Sub Button1_MouseClick(sender As Object, e As MouseEventArgs) Handles Button1.MouseClick
        'Start the timer
        Me.Timer1.Start()
    End Sub

    Private Sub Button2_MouseClick(sender As Object, e As MouseEventArgs) Handles Button2.MouseClick
        'Stop the timer
        Me.Timer1.Stop()
    End Sub

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Me.RichTextBox1.AppendText("Exam".PadRight(10) & "Grade".PadRight(8) & "Current Mean" & Environment.NewLine)

        Me.Chart1.ChartAreas("ChartArea1").AxisX.MinorTickMark.Enabled = True
        Me.Chart1.ChartAreas("ChartArea1").AxisX.Interval = 1
        Me.Chart1.ChartAreas("ChartArea1").AxisX.IsLabelAutoFit = True

    End Sub

End Class

'Class frequencies
Public Class frequencies

    Public countFrequencies As Integer = 1
    Public RelativeFrequencies As Double
    Public PercentageFrequencies As Double

End Class

